File: server/camcops_server/tasks/qolbasic.py
#!/usr/bin/env python
# camcops_server/tasks/qolbasic.py

"""
===============================================================================

    Copyright (C) 2012-2018 Rudolf Cardinal (rudolf@pobox.com).

    This file is part of CamCOPS.

    CamCOPS is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    CamCOPS is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with CamCOPS. If not, see <http://www.gnu.org/licenses/>.

===============================================================================
"""

from typing import List, Optional

from cardinal_pythonlib.maths_py import mean
import cardinal_pythonlib.rnc_web as ws
from sqlalchemy.sql.sqltypes import Float

from camcops_server.cc_modules.cc_constants import CssClass
from camcops_server.cc_modules.cc_ctvinfo import CTV_INCOMPLETE, CtvInfo
from camcops_server.cc_modules.cc_html import answer, identity, tr
from camcops_server.cc_modules.cc_request import CamcopsRequest
from camcops_server.cc_modules.cc_sqla_coltypes import (
    CamcopsColumn,
    PermittedValueChecker,
)
from camcops_server.cc_modules.cc_summaryelement import SummaryElement
from camcops_server.cc_modules.cc_task import Task, TaskHasPatientMixin
from camcops_server.cc_modules.cc_trackerhelpers import TrackerInfo


# =============================================================================
# QoL-Basic
# =============================================================================

DP = 3


class QolBasic(TaskHasPatientMixin, Task):
    __tablename__ = "qolbasic"
    shortname = "QoL-Basic"
    longname = "Quality of Life: basic assessment"
    provides_trackers = True

    tto = CamcopsColumn(
        "tto", Float,
        permitted_value_checker=PermittedValueChecker(minimum=0, maximum=10),
        comment="Time trade-off (QoL * 10). Prompt: ... Indicate... the "
                "number of years in full health [0-10] that you think is "
                "of equal value to 10 years in your current health state."
    )
    rs = CamcopsColumn(
        "rs", Float,
        permitted_value_checker=PermittedValueChecker(minimum=0, maximum=100),
        comment="Rating scale (QoL * 100). Prompt: Mark the point on the "
                "scale [0-100] that you feel best illustrates your current "
                "quality of life."
    )

    TASK_FIELDS = ["tto", "rs"]

    def get_trackers(self, req: CamcopsRequest) -> List[TrackerInfo]:
        return [
            TrackerInfo(
                value=self.get_tto_qol(),
                plot_label="Quality of life: time trade-off",
                axis_label="TTO QoL (0-1)",
                axis_min=0,
                axis_max=1
            ),
            TrackerInfo(
                value=self.get_rs_qol(),
                plot_label="Quality of life: rating scale",
                axis_label="RS QoL (0-1)",
                axis_min=0,
                axis_max=1
            ),
        ]

    def get_clinical_text(self, req: CamcopsRequest) -> List[CtvInfo]:
        if not self.is_complete():
            return CTV_INCOMPLETE
        tto_qol = self.get_tto_qol()
        rs_qol = self.get_rs_qol()
        mean_qol = mean([tto_qol, rs_qol])
        return [CtvInfo(
            content=(
                "Quality of life: time trade-off {}, rating scale {}, "
                "mean {}.".format(
                    ws.number_to_dp(tto_qol, DP),
                    ws.number_to_dp(rs_qol, DP),
                    ws.number_to_dp(mean_qol, DP)
                )
            )
        )]

    def get_summaries(self, req: CamcopsRequest) -> List[SummaryElement]:
        return self.standard_task_summary_fields() + [
            SummaryElement(
                name="tto_qol", coltype=Float(),
                value=self.get_tto_qol(),
                comment="Quality of life (0-1), from time trade-off method"),
            SummaryElement(
                name="rs_qol", coltype=Float(),
                value=self.get_rs_qol(),
                comment="Quality of life (0-1), from rating scale method"),
        ]

    def is_complete(self) -> bool:
        return (
            self.are_all_fields_complete(QolBasic.TASK_FIELDS) and
            self.field_contents_valid()
        )

    def get_tto_qol(self) -> Optional[float]:
        return self.tto / 10 if self.tto is not None else None

    def get_rs_qol(self) -> Optional[float]:
        return self.rs / 100 if self.rs is not None else None

    def get_task_html(self, req: CamcopsRequest) -> str:
        tto_qol = self.get_tto_qol()
        rs_qol = self.get_rs_qol()
        mean_qol = mean([tto_qol, rs_qol])
        h = """
            <div class="{CssClass.SUMMARY}">
                <table class="{CssClass.SUMMARY}">
                    {tr_is_complete}
                    {mean_qol}
                </table>
            </div>
            <div class="{CssClass.EXPLANATION}">
                Quality of life (QoL) has anchor values of 0 (none) and 1
                (perfect health), and can be asked about in several ways.
            </div>
            <table class="{CssClass.TASKDETAIL}">
                <tr>
                    <th width="33%">Scale</th>
                    <th width="33%">Answer</th>
                    <td width="33%">Implied QoL</th>
                </tr>
                {tto}
                {rs}
            </table>
        """.format(
            CssClass=CssClass,
            tr_is_complete=self.get_is_complete_tr(req),
            mean_qol=tr(
                "Mean QoL",
                answer(ws.number_to_dp(mean_qol, DP, default=None),
                       formatter_answer=identity)
            ),
            tto=tr(
                self.wxstring(req, "tto_q_s"),
                answer(ws.number_to_dp(self.tto, DP, default=None)),
                answer(ws.number_to_dp(tto_qol, DP, default=None),
                       formatter_answer=identity)
            ),
            rs=tr(
                self.wxstring(req, "rs_q_s"),
                answer(ws.number_to_dp(self.rs, DP, default=None)),
                answer(ws.number_to_dp(rs_qol, DP, default=None),
                       formatter_answer=identity)
            ),
        )
        return h

File: tablet_qt/tasks/qolbasic.cpp
/*
    Copyright (C) 2012-2018 Rudolf Cardinal (rudolf@pobox.com).

    This file is part of CamCOPS.

    CamCOPS is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    CamCOPS is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with CamCOPS. If not, see <http://www.gnu.org/licenses/>.
*/

#include "qolbasic.h"
#include "lib/convert.h"
#include "maths/mathfunc.h"
#include "lib/stringfunc.h"
#include "questionnairelib/namevaluepair.h"
#include "questionnairelib/questionnaire.h"
#include "questionnairelib/quslider.h"
#include "questionnairelib/qutext.h"
#include "tasklib/taskfactory.h"
using mathfunc::mean;
using mathfunc::noneNull;
using stringfunc::standardResult;


const QString QolBasic::QOLBASIC_TABLENAME("qolbasic");
const QString TTO("tto");
const QString RS("rs");

const int DP_TTO = 2;
const int DP_RS = 2;
const int DP_MEAN = 3;


void initializeQolBasic(TaskFactory& factory)
{
    static TaskRegistrar<QolBasic> registered(factory);
}


QolBasic::QolBasic(CamcopsApp& app, DatabaseManager& db, const int load_pk) :
    Task(app, db, QOLBASIC_TABLENAME, false, false, false)  // ... anon, clin, resp
{
    addField(TTO, QVariant::Double);
    addField(RS, QVariant::Double);

    load(load_pk);  // MUST ALWAYS CALL from derived Task constructor.
}


// ============================================================================
// Class info
// ============================================================================

QString QolBasic::shortname() const
{
    return "QoL-Basic";
}


QString QolBasic::longname() const
{
    return tr("Quality of Life: basic assessment");
}


QString QolBasic::menusubtitle() const
{
    return tr("Time trade-off and response scale measures of quality of life.");
}


QString QolBasic::infoFilenameStem() const
{
    return "qol";
}


// ============================================================================
// Instance info
// ============================================================================

bool QolBasic::isComplete() const
{
    return noneNull(values({TTO, RS}));
}


QStringList QolBasic::summary() const
{
    return QStringList{
        standardResult(xstring("tto_q_s"),
                       convert::prettyValue(qolTto(), DP_TTO)),
        standardResult(xstring("rs_q_s"),
                       convert::prettyValue(qolRs(), DP_RS)),
        standardResult(xstring("mean_qol"),
                       convert::prettyValue(meanQol(), DP_MEAN)),
    };
}


QStringList QolBasic::detail() const
{
    return completenessInfo() + summary();
}


OpenableWidget* QolBasic::editor(const bool read_only)
{
    // The TTO slider goes from 0-10 in steps of 0.1
    // QuSlider uses integers internally but can scale.
    // So we say:
    // (a) use 0-100 in steps of 1 internally
    // (b) scale for real output of 0-10 with 1dp
    QuSlider* tto_slider = new QuSlider(fieldRef(TTO), 0, 100, 1);
    tto_slider->setConvertForRealField(true, 0, 10.0, 1);
    tto_slider->setBigStep(10);
    tto_slider->setTickInterval(10);
    tto_slider->setNullApparentValue(0);  // debatable: 0? 50?
    tto_slider->setHorizontal(true);
    tto_slider->setShowValue(true);
    tto_slider->setTickLabels({
        {0, "0"},
        {10, "1"},
        {20, "2"},
        {30, "3"},
        {40, "4"},
        {50, "5"},
        {60, "6"},
        {70, "7"},
        {80, "8"},
        {90, "9"},
        {100, "10"},
    });
    tto_slider->setEdgeInExtremeLabels(false);
    tto_slider->setTickPosition(QSlider::TicksBothSides);
    tto_slider->setTickLabelPosition(QSlider::TicksBelow);

    // The RS slider goes from 0-100 in steps of 1.
    QuSlider* rs_slider = new QuSlider(fieldRef(RS), 0, 100, 1);
    rs_slider->setBigStep(10);
    rs_slider->setTickInterval(10);
    rs_slider->setNullApparentValue(0);  // debatable: 0? 50?
    rs_slider->setHorizontal(true);
    rs_slider->setShowValue(true);
    rs_slider->setTickLabels({
        {0, xstring("rs_0")},
        {100, xstring("rs_100")},
    });
    rs_slider->setTickInterval(100);  // just 0 and 100
    rs_slider->setEdgeInExtremeLabels(true);
    rs_slider->setTickPosition(QSlider::TicksBothSides);
    rs_slider->setTickLabelPosition(QSlider::TicksBelow);

    QuPagePtr page1((new QuPage{
        new QuText(xstring("tto_q")),
        tto_slider,
    })->setTitle(xstring("tto_title")));

    QuPagePtr page2((new QuPage{
         new QuText(xstring("rs_q")),
         rs_slider,
    })->setTitle(xstring("rs_title")));

    Questionnaire* questionnaire = new Questionnaire(m_app, {page1, page2});
    questionnaire->setType(QuPage::PageType::Patient);
    questionnaire->setReadOnly(read_only);
    return questionnaire;
}


// ============================================================================
// Task-specific calculations
// ============================================================================

QVariant QolBasic::qolTto() const
{
    const QVariant tto = value(TTO);
    return tto.isNull() ? tto : tto.toDouble() / 10.0;
}


QVariant QolBasic::qolRs() const
{
    const QVariant rs = value(RS);
    return rs.isNull() ? rs : rs.toDouble() / 100.0;
}


QVariant QolBasic::meanQol() const
{
    const QVariant tto = qolTto();
    const QVariant rs = qolRs();
    return mean({tto, rs}, true);
}
